File: src/main/java/no/nav/sosialhjelp/soknad/business/service/oppsummering/steg/FamiliesituasjonSteg.java
package no.nav.sosialhjelp.soknad.business.service.oppsummering.steg;

import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonAnsvar;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonBarnebidrag;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonEktefelle;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonForsorgerplikt;
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonSivilstatus;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Avsnitt;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Felt;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Sporsmal;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Steg;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Svar;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.SvarType;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Type;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

import static java.util.Collections.singletonList;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.StegUtils.booleanVerdiFelt;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.StegUtils.createSvar;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.StegUtils.fulltnavn;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.StegUtils.harBarnMedKilde;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.StegUtils.isNotNullOrEmtpy;

public class FamiliesituasjonSteg {

    public Steg get(JsonInternalSoknad jsonInternalSoknad) {
        var familie = jsonInternalSoknad.getSoknad().getData().getFamilie();

        return new Steg.Builder()
                .withStegNr(4)
                .withTittel("familiebolk.tittel")
                .withAvsnitt(
                        List.of(
                                sivilstatusAvsnitt(familie.getSivilstatus()),
                                forsorgerpliktAvsnitt(familie.getForsorgerplikt())
                        )
                )
                .build();
    }

    private Avsnitt sivilstatusAvsnitt(JsonSivilstatus sivilstatus) {
        return new Avsnitt.Builder()
                .withTittel("system.familie.sivilstatus.sporsmal")
                .withSporsmal(sivilstatusSporsmal(sivilstatus))
                .build();
    }

    private List<Sporsmal> sivilstatusSporsmal(JsonSivilstatus sivilstatus) {
        var harUtfyltSivilstatusSporsmal = sivilstatus != null && sivilstatus.getKilde() != null;
        var harSystemEktefelle = harUtfyltSivilstatusSporsmal && sivilstatus.getKilde().equals(JsonKilde.SYSTEM) && sivilstatus.getStatus().equals(JsonSivilstatus.Status.GIFT);
        var harSystemEktefelleMedAdressebeskyttelse = harSystemEktefelle && Boolean.TRUE.equals(sivilstatus.getEktefelleHarDiskresjonskode());
        var harBrukerUtfyltSivilstatus = harUtfyltSivilstatusSporsmal && !harSystemEktefelle && sivilstatus.getKilde().equals(JsonKilde.BRUKER);
        var harBrukerUtfyltEktefelle = harBrukerUtfyltSivilstatus && sivilstatus.getStatus().equals(JsonSivilstatus.Status.GIFT) && sivilstatus.getEktefelle() != null;

        var sporsmal = new ArrayList<Sporsmal>();

        if (!harUtfyltSivilstatusSporsmal) {
            sporsmal.add(brukerSivilstatusSporsmal(false, null));
        }

        if (harBrukerUtfyltSivilstatus && !harBrukerUtfyltEktefelle) {
            sporsmal.add(brukerSivilstatusSporsmal(true, sivilstatus.getStatus()));
        }

        if (harBrukerUtfyltEktefelle) {
            sporsmal.add(brukerRegistrertEktefelle(sivilstatus));
        }

        if (harSystemEktefelleMedAdressebeskyttelse) {
            sporsmal.add(systemEktefelleMedAdressebeskyttelseSporsmal());
        }

        if (harSystemEktefelle && !harSystemEktefelleMedAdressebeskyttelse) {
            sporsmal.add(systemEktefelleSporsmal(sivilstatus));
        }

        return sporsmal;
    }

    private Sporsmal brukerSivilstatusSporsmal(boolean erUtfylt, JsonSivilstatus.Status status) {
        return new Sporsmal.Builder()
                .withTittel("familie.sivilstatus.sporsmal")
                .withErUtfylt(erUtfylt)
                .withFelt(erUtfylt ?
                        singletonList(
                                new Felt.Builder()
                                        .withSvar(createSvar(statusToTekstKey(status), SvarType.LOCALE_TEKST))
                                        .withType(Type.CHECKBOX)
                                        .build()
                        ) :
                        null
                )
                .build();
    }

    private String statusToTekstKey(JsonSivilstatus.Status status) {
        String key;
        switch (status) {
            case ENKE:
                key = "familie.sivilstatus.enke";
                break;
            case GIFT:
                key = "familie.sivilstatus.gift";
                break;
            case SKILT:
                key = "familie.sivilstatus.skilt";
                break;
            case SAMBOER:
                key = "familie.sivilstatus.samboer";
                break;
            case SEPARERT:
                key = "familie.sivilstatus.separert";
                break;
            case UGIFT:
            default:
                key = "familie.sivilstatus.ugift";
                break;
        }
        return key;
    }

    private Sporsmal brukerRegistrertEktefelle(JsonSivilstatus sivilstatus) {
        // todo: som dette eller som liste av sporsmal?
        var ektefelle = sivilstatus.getEktefelle();
        var erUtfylt = isNotNullOrEmtpy(ektefelle.getNavn().getFornavn()) &&
                isNotNullOrEmtpy(ektefelle.getNavn().getEtternavn()) &&
                isNotNullOrEmtpy(ektefelle.getFodselsdato()) &&
                isNotNullOrEmtpy(ektefelle.getPersonIdentifikator()) &&
                sivilstatus.getBorSammenMed() != null;

        var map = new LinkedHashMap<String, Svar>();
        map.put("familie.sivilstatus.gift.ektefelle.navn.label", createSvar(fulltnavn(ektefelle.getNavn()), SvarType.TEKST));
        map.put("familie.sivilstatus.gift.ektefelle.fnr.label", createSvar(ektefelle.getFodselsdato(), SvarType.DATO));
        map.put("familie.sivilstatus.gift.ektefelle.pnr.label", createSvar(personnummerFraFnr(ektefelle), SvarType.TEKST));
        map.put("familie.sivilstatus.gift.ektefelle.borsammen.sporsmal", createSvar(borSammenMedSvar(sivilstatus), SvarType.LOCALE_TEKST));

        return new Sporsmal.Builder()
                .withTittel("familie.sivilstatus.gift.ektefelle.sporsmal")
                .withErUtfylt(erUtfylt)
                .withFelt(singletonList(
                        new Felt.Builder()
                                .withType(Type.SYSTEMDATA_MAP) //selv om dette ikke er systemdata?!?
                                .withLabelSvarMap(map)
                                .build()
                ))
                .build();
    }

    private String personnummerFraFnr(JsonEktefelle ektefelle) {
        if (ektefelle.getPersonIdentifikator() != null && ektefelle.getPersonIdentifikator().length() == 11) {
            return ektefelle.getPersonIdentifikator().substring(6, 11);
        }
        return ektefelle.getPersonIdentifikator();
    }

    private String borSammenMedSvar(JsonSivilstatus sivilstatus) {
        return sivilstatus.getBorSammenMed() != null ? borSammenKey(sivilstatus) : null;
    }

    private String borSammenKey(JsonSivilstatus sivilstatus) {
        return Boolean.TRUE.equals(sivilstatus.getBorSammenMed()) ? "familie.sivilstatus.gift.ektefelle.borsammen.true" : "familie.sivilstatus.gift.ektefelle.borsammen.false";
    }

    private Sporsmal systemEktefelleMedAdressebeskyttelseSporsmal() {
        return new Sporsmal.Builder()
                .withTittel("system.familie.sivilstatus")
                .withErUtfylt(true)
                .withFelt(
                        singletonList(
                                new Felt.Builder()
                                        .withSvar(createSvar("system.familie.sivilstatus.ikkeTilgang.label", SvarType.LOCALE_TEKST))
                                        .withType(Type.SYSTEMDATA)
                                        .build()
                        )
                )
                .build();
    }

    private Sporsmal systemEktefelleSporsmal(JsonSivilstatus sivilstatus) {
        var ektefelle = sivilstatus.getEktefelle();
        var labelSvarMap = new LinkedHashMap<String, Svar>();
        if (ektefelle.getNavn() != null) {
            labelSvarMap.put("system.familie.sivilstatus.gift.ektefelle.navn", createSvar(fulltnavn(ektefelle.getNavn()), SvarType.TEKST));
        }
        if (ektefelle.getFodselsdato() != null) {
            labelSvarMap.put("system.familie.sivilstatus.gift.ektefelle.fodselsdato", createSvar(ektefelle.getFodselsdato(), SvarType.DATO));
        }
        if (sivilstatus.getFolkeregistrertMedEktefelle() != null) {
            labelSvarMap.put("system.familie.sivilstatus.gift.ektefelle.folkereg", createSvar(Boolean.TRUE.equals(sivilstatus.getFolkeregistrertMedEktefelle()) ? "system.familie.sivilstatus.gift.ektefelle.folkeregistrertsammen.true" : "system.familie.sivilstatus.gift.ektefelle.folkeregistrertsammen.false", SvarType.LOCALE_TEKST));
        }

        return new Sporsmal.Builder()
                .withTittel("system.familie.sivilstatus.infotekst")
                .withErUtfylt(true)
                .withFelt(
                        singletonList(
                                new Felt.Builder()
                                        .withType(Type.SYSTEMDATA_MAP)
                                        .withLabelSvarMap(labelSvarMap)
                                        .build()
                        )
                )
                .build();
    }

    private Avsnitt forsorgerpliktAvsnitt(JsonForsorgerplikt forsorgerplikt) {
        return new Avsnitt.Builder()
                .withTittel("familierelasjon.faktum.sporsmal")
                .withSporsmal(forsorgerpliktSporsmal(forsorgerplikt))
                .build();
    }

    private List<Sporsmal> forsorgerpliktSporsmal(JsonForsorgerplikt forsorgerplikt) {
        var harSystemBarn = harBarnMedKilde(forsorgerplikt, JsonKilde.SYSTEM);
        var harBrukerBarn = harBarnMedKilde(forsorgerplikt, JsonKilde.BRUKER);

        var sporsmal = new ArrayList<Sporsmal>();

        if (!harSystemBarn && !harBrukerBarn) {
            sporsmal.add(ingenRegistrerteBarnSporsmal());
        }

        if (harSystemBarn) {
            forsorgerplikt.getAnsvar().stream()
                    .filter(barn -> barn.getBarn().getKilde().equals(JsonKilde.SYSTEM))
                    .forEach(barn -> {
                                sporsmal.add(systemBarnSporsmal(barn));
                                if (Boolean.TRUE.equals(barn.getErFolkeregistrertSammen().getVerdi())) {
                                    sporsmal.add(deltBostedSporsmal(barn));
                                }
                            }
                    );
        }

        if (harBrukerBarn) {
            // todo brukerregistrerte barn. pt ikke støtte for dette i søknad
        }

        if (harSystemBarn || harBrukerBarn) {
            sporsmal.add(barneBidragSporsmal(forsorgerplikt));
        }

        return sporsmal;
    }

    private Sporsmal ingenRegistrerteBarnSporsmal() {
        return new Sporsmal.Builder()
                .withTittel("familierelasjon.ingen_registrerte_barn_tittel")
                .withErUtfylt(true)
                .withFelt(
                        singletonList(
                                new Felt.Builder()
                                        .withSvar(createSvar("familierelasjon.ingen_registrerte_barn_tekst", SvarType.LOCALE_TEKST))
                                        .withType(Type.SYSTEMDATA)
                                        .build()
                        )
                )
                .build();
    }

    private Sporsmal systemBarnSporsmal(JsonAnsvar barn) {
        var labelSvarMap = new LinkedHashMap<String, Svar>();
        if (barn.getBarn().getNavn() != null) {
            labelSvarMap.put("familie.barn.true.barn.navn.label", createSvar(fulltnavn(barn.getBarn().getNavn()), SvarType.TEKST));
        }
        if (barn.getBarn().getFodselsdato() != null) {
            labelSvarMap.put("familierelasjon.fodselsdato", createSvar(barn.getBarn().getFodselsdato(), SvarType.DATO));
        }
        if (barn.getErFolkeregistrertSammen() != null) {
            labelSvarMap.put("familierelasjon.samme_folkeregistrerte_adresse", createSvar(Boolean.TRUE.equals(barn.getErFolkeregistrertSammen().getVerdi()) ? "system.familie.barn.true.barn.folkeregistrertsammen.true" : "system.familie.barn.true.barn.folkeregistrertsammen.false", SvarType.LOCALE_TEKST));
        }

        return new Sporsmal.Builder()
                .withTittel("familie.barn.true.barn.sporsmal")
                .withErUtfylt(true)
                .withFelt(
                        singletonList(
                                new Felt.Builder()
                                        .withType(Type.SYSTEMDATA_MAP)
                                        .withLabelSvarMap(labelSvarMap)
                                        .build()
                        )
                )
                .build();
    }

    private Sporsmal deltBostedSporsmal(JsonAnsvar barn) {
        var harUtfyltDeltBostedSporsmal = barn.getHarDeltBosted() != null && barn.getHarDeltBosted().getVerdi() != null;
        var harSvartJaDeltBosted = harUtfyltDeltBostedSporsmal && Boolean.TRUE.equals(barn.getHarDeltBosted().getVerdi());

        return new Sporsmal.Builder()
                .withTittel("system.familie.barn.true.barn.deltbosted.sporsmal")
                .withErUtfylt(harUtfyltDeltBostedSporsmal)
                .withFelt(harUtfyltDeltBostedSporsmal ?
                        booleanVerdiFelt(harSvartJaDeltBosted, "system.familie.barn.true.barn.deltbosted.true", "system.familie.barn.true.barn.deltbosted.false") :
                        null
                )
                .build();

    }

    private Sporsmal barneBidragSporsmal(JsonForsorgerplikt forsorgerplikt) {
        var erUtfylt = forsorgerplikt.getBarnebidrag() != null && forsorgerplikt.getBarnebidrag().getVerdi() != null;
        return new Sporsmal.Builder()
                .withTittel("familie.barn.true.barnebidrag.sporsmal")
                .withErUtfylt(erUtfylt)
                .withFelt(erUtfylt ?
                        singletonList(
                                new Felt.Builder()
                                        .withSvar(createSvar(verdiToTekstKey(forsorgerplikt.getBarnebidrag().getVerdi()), SvarType.LOCALE_TEKST))
                                        .withType(Type.CHECKBOX)
                                        .build()
                        ) :
                        null
                )
                .build();
    }

    private String verdiToTekstKey(JsonBarnebidrag.Verdi verdi) {
        String key;
        switch (verdi) {
            case MOTTAR:
                key = "familie.barn.true.barnebidrag.mottar";
                break;
            case BETALER:
                key = "familie.barn.true.barnebidrag.betaler";
                break;
            case BEGGE:
                key = "familie.barn.true.barnebidrag.begge";
                break;
            case INGEN:
            default:
                key = "familie.barn.true.barnebidrag.ingen";
                break;
        }
        return key;
    }

}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/oppsummering/steg/FamiliesituasjonSteg.kt
package no.nav.sosialhjelp.soknad.oppsummering.steg

import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonAnsvar
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonBarnebidrag.Verdi
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonEktefelle
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonForsorgerplikt
import no.nav.sbl.soknadsosialhjelp.soknad.familie.JsonSivilstatus
import no.nav.sosialhjelp.soknad.oppsummering.dto.Avsnitt
import no.nav.sosialhjelp.soknad.oppsummering.dto.Felt
import no.nav.sosialhjelp.soknad.oppsummering.dto.Sporsmal
import no.nav.sosialhjelp.soknad.oppsummering.dto.Steg
import no.nav.sosialhjelp.soknad.oppsummering.dto.Svar
import no.nav.sosialhjelp.soknad.oppsummering.dto.SvarType
import no.nav.sosialhjelp.soknad.oppsummering.dto.Type
import no.nav.sosialhjelp.soknad.oppsummering.steg.StegUtils.createSvar
import no.nav.sosialhjelp.soknad.oppsummering.steg.StegUtils.fulltnavn
import no.nav.sosialhjelp.soknad.oppsummering.steg.StegUtils.harBrukerRegistrerteBarn
import no.nav.sosialhjelp.soknad.oppsummering.steg.StegUtils.harSystemRegistrerteBarn
import no.nav.sosialhjelp.soknad.oppsummering.steg.StegUtils.isNotNullOrEmtpy

class FamiliesituasjonSteg {

    fun get(jsonInternalSoknad: JsonInternalSoknad): Steg {
        val familie = jsonInternalSoknad.soknad.data.familie
        return Steg(
            stegNr = 4,
            tittel = "familiebolk.tittel",
            avsnitt = listOf(
                sivilstatusAvsnitt(familie.sivilstatus),
                forsorgerpliktAvsnitt(familie.forsorgerplikt)
            )
        )
    }

    private fun sivilstatusAvsnitt(sivilstatus: JsonSivilstatus?): Avsnitt {
        return Avsnitt(
            tittel = "system.familie.sivilstatus.sporsmal",
            sporsmal = sivilstatusSporsmal(sivilstatus)
        )
    }

    private fun sivilstatusSporsmal(sivilstatus: JsonSivilstatus?): List<Sporsmal> {
        val harUtfyltSivilstatusSporsmal = sivilstatus != null && sivilstatus.kilde != null
        val harSystemEktefelle =
            harUtfyltSivilstatusSporsmal && sivilstatus!!.kilde == JsonKilde.SYSTEM && sivilstatus.status == JsonSivilstatus.Status.GIFT
        val harSystemEktefelleMedAdressebeskyttelse =
            harSystemEktefelle && java.lang.Boolean.TRUE == sivilstatus!!.ektefelleHarDiskresjonskode
        val harBrukerUtfyltSivilstatus =
            harUtfyltSivilstatusSporsmal && !harSystemEktefelle && sivilstatus!!.kilde == JsonKilde.BRUKER
        val harBrukerUtfyltEktefelle =
            harBrukerUtfyltSivilstatus && sivilstatus!!.status == JsonSivilstatus.Status.GIFT && sivilstatus.ektefelle != null
        val sporsmal = ArrayList<Sporsmal>()
        if (!harUtfyltSivilstatusSporsmal) {
            sporsmal.add(brukerSivilstatusSporsmal(false, null))
        }
        if (harBrukerUtfyltSivilstatus && !harBrukerUtfyltEktefelle) {
            sporsmal.add(brukerSivilstatusSporsmal(true, sivilstatus!!.status))
        }
        if (harBrukerUtfyltEktefelle) {
            sporsmal.add(brukerRegistrertEktefelle(sivilstatus))
        }
        if (harSystemEktefelleMedAdressebeskyttelse) {
            sporsmal.add(systemEktefelleMedAdressebeskyttelseSporsmal())
        }
        if (harSystemEktefelle && !harSystemEktefelleMedAdressebeskyttelse) {
            sporsmal.add(systemEktefelleSporsmal(sivilstatus))
        }
        return sporsmal
    }

    private fun brukerSivilstatusSporsmal(erUtfylt: Boolean, status: JsonSivilstatus.Status?): Sporsmal {
        return Sporsmal(
            tittel = "familie.sivilstatus.sporsmal",
            erUtfylt = erUtfylt,
            felt = if (erUtfylt) listOf(
                Felt(
                    type = Type.CHECKBOX,
                    svar = createSvar(statusToTekstKey(status), SvarType.LOCALE_TEKST)
                )
            ) else null
        )
    }

    private fun statusToTekstKey(status: JsonSivilstatus.Status?): String {
        val key: String = when (status) {
            JsonSivilstatus.Status.ENKE -> "familie.sivilstatus.enke"
            JsonSivilstatus.Status.GIFT -> "familie.sivilstatus.gift"
            JsonSivilstatus.Status.SKILT -> "familie.sivilstatus.skilt"
            JsonSivilstatus.Status.SAMBOER -> "familie.sivilstatus.samboer"
            JsonSivilstatus.Status.SEPARERT -> "familie.sivilstatus.separert"
            JsonSivilstatus.Status.UGIFT -> "familie.sivilstatus.ugift"
            else -> "familie.sivilstatus.ugift"
        }
        return key
    }

    private fun brukerRegistrertEktefelle(sivilstatus: JsonSivilstatus?): Sporsmal {
        // todo: som dette eller som liste av sporsmal?
        val ektefelle = sivilstatus!!.ektefelle
        val erUtfylt = isNotNullOrEmtpy(ektefelle.navn.fornavn) &&
            isNotNullOrEmtpy(ektefelle.navn.etternavn) &&
            isNotNullOrEmtpy(ektefelle.fodselsdato) &&
            isNotNullOrEmtpy(ektefelle.personIdentifikator) && sivilstatus.borSammenMed != null
        val map = LinkedHashMap<String, Svar>()
        map["familie.sivilstatus.gift.ektefelle.navn.label"] =
            createSvar(fulltnavn(ektefelle.navn), SvarType.TEKST)
        map["familie.sivilstatus.gift.ektefelle.fnr.label"] =
            createSvar(ektefelle.fodselsdato, SvarType.DATO)
        map["familie.sivilstatus.gift.ektefelle.pnr.label"] =
            createSvar(personnummerFraFnr(ektefelle), SvarType.TEKST)
        map["familie.sivilstatus.gift.ektefelle.borsammen.sporsmal"] =
            createSvar(borSammenMedSvar(sivilstatus), SvarType.LOCALE_TEKST)
        return Sporsmal(
            tittel = "familie.sivilstatus.gift.ektefelle.sporsmal",
            erUtfylt = erUtfylt,
            felt = listOf(
                Felt(
                    type = Type.SYSTEMDATA_MAP, // selv om dette ikke er systemdata?!?
                    labelSvarMap = map
                )
            )
        )
    }

    private fun personnummerFraFnr(ektefelle: JsonEktefelle): String? {
        return if (ektefelle.personIdentifikator != null && ektefelle.personIdentifikator.length == 11) {
            ektefelle.personIdentifikator.substring(6, 11)
        } else ektefelle.personIdentifikator
    }

    private fun borSammenMedSvar(sivilstatus: JsonSivilstatus?): String? {
        return if (sivilstatus!!.borSammenMed != null) borSammenKey(sivilstatus) else null
    }

    private fun borSammenKey(sivilstatus: JsonSivilstatus?): String {
        return if (java.lang.Boolean.TRUE == sivilstatus!!.borSammenMed) "familie.sivilstatus.gift.ektefelle.borsammen.true" else "familie.sivilstatus.gift.ektefelle.borsammen.false"
    }

    private fun systemEktefelleMedAdressebeskyttelseSporsmal(): Sporsmal {
        return Sporsmal(
            tittel = "system.familie.sivilstatus",
            erUtfylt = true,
            felt = listOf(
                Felt(
                    type = Type.SYSTEMDATA,
                    svar = createSvar("system.familie.sivilstatus.ikkeTilgang.label", SvarType.LOCALE_TEKST)
                )
            )
        )
    }

    private fun systemEktefelleSporsmal(sivilstatus: JsonSivilstatus?): Sporsmal {
        val ektefelle = sivilstatus!!.ektefelle
        val labelSvarMap = LinkedHashMap<String, Svar>()
        if (ektefelle.navn != null) {
            labelSvarMap["system.familie.sivilstatus.gift.ektefelle.navn"] =
                createSvar(fulltnavn(ektefelle.navn), SvarType.TEKST)
        }
        if (ektefelle.fodselsdato != null) {
            labelSvarMap["system.familie.sivilstatus.gift.ektefelle.fodselsdato"] =
                createSvar(ektefelle.fodselsdato, SvarType.DATO)
        }
        if (sivilstatus.folkeregistrertMedEktefelle != null) {
            labelSvarMap["system.familie.sivilstatus.gift.ektefelle.folkereg"] = createSvar(
                if (java.lang.Boolean.TRUE == sivilstatus.folkeregistrertMedEktefelle) "system.familie.sivilstatus.gift.ektefelle.folkeregistrertsammen.true" else "system.familie.sivilstatus.gift.ektefelle.folkeregistrertsammen.false",
                SvarType.LOCALE_TEKST
            )
        }
        return Sporsmal(
            tittel = "system.familie.sivilstatus.infotekst",
            erUtfylt = true,
            felt = listOf(
                Felt(
                    type = Type.SYSTEMDATA_MAP,
                    labelSvarMap = labelSvarMap
                )
            )
        )
    }

    private fun forsorgerpliktAvsnitt(forsorgerplikt: JsonForsorgerplikt?): Avsnitt {
        return Avsnitt(
            tittel = "familierelasjon.faktum.sporsmal",
            sporsmal = forsorgerpliktSporsmal(forsorgerplikt)
        )
    }

    private fun forsorgerpliktSporsmal(forsorgerplikt: JsonForsorgerplikt?): List<Sporsmal> {
        val harSystemBarn = harSystemRegistrerteBarn(forsorgerplikt)
        val harBrukerBarn = harBrukerRegistrerteBarn(forsorgerplikt)
        val sporsmal = ArrayList<Sporsmal>()
        if (!harSystemBarn && !harBrukerBarn) {
            sporsmal.add(ingenRegistrerteBarnSporsmal())
        }
        if (harSystemBarn) {
            forsorgerplikt?.ansvar
                ?.filter { it.barn.kilde == JsonKilde.SYSTEM }
                ?.forEach { barn: JsonAnsvar ->
                    sporsmal.add(systemBarnSporsmal(barn))
                    if (java.lang.Boolean.TRUE == barn.erFolkeregistrertSammen.verdi) {
                        sporsmal.add(deltBostedSporsmal(barn))
                    }
                }
        }
        if (harBrukerBarn) {
            forsorgerplikt?.ansvar
                ?.filter { it.barn.kilde == JsonKilde.BRUKER }
                ?.forEach {
                    sporsmal.add(brukerBarnSporsmal(it))
                    if (it.borSammenMed != null && java.lang.Boolean.TRUE == it.borSammenMed.verdi) {
                        sporsmal.add(deltBostedSporsmal(it))
                    }
                }
        }
        if (harSystemBarn || harBrukerBarn) {
            sporsmal.add(barneBidragSporsmal(forsorgerplikt))
        }
        return sporsmal
    }

    private fun ingenRegistrerteBarnSporsmal(): Sporsmal {
        return Sporsmal(
            tittel = "familierelasjon.ingen_registrerte_barn_tittel",
            erUtfylt = true,
            felt = listOf(
                Felt(
                    type = Type.SYSTEMDATA,
                    svar = createSvar("familierelasjon.ingen_registrerte_barn_tekst", SvarType.LOCALE_TEKST)
                )
            )
        )
    }

    private fun systemBarnSporsmal(barn: JsonAnsvar): Sporsmal {
        return barnSporsmal(barn, erSystemRegistrert = true)
    }

    private fun brukerBarnSporsmal(barn: JsonAnsvar): Sporsmal {
        return barnSporsmal(barn, erSystemRegistrert = false)
    }

    private fun barnSporsmal(barn: JsonAnsvar, erSystemRegistrert: Boolean): Sporsmal {
        val labelSvarMap = LinkedHashMap<String, Svar>()
        if (barn.barn.navn != null) {
            labelSvarMap["familie.barn.true.barn.navn.label"] = createSvar(fulltnavn(barn.barn.navn), SvarType.TEKST)
        }
        if (barn.barn.fodselsdato != null) {
            labelSvarMap["familierelasjon.fodselsdato"] = createSvar(barn.barn.fodselsdato, SvarType.DATO)
        }
        if (erSystemRegistrert && barn.erFolkeregistrertSammen != null) {
            labelSvarMap["familierelasjon.samme_folkeregistrerte_adresse"] = createSvar(
                if (java.lang.Boolean.TRUE == barn.erFolkeregistrertSammen.verdi) "system.familie.barn.true.barn.folkeregistrertsammen.true" else "system.familie.barn.true.barn.folkeregistrertsammen.false",
                SvarType.LOCALE_TEKST
            )
        }
        if (!erSystemRegistrert && barn.borSammenMed != null) {
            labelSvarMap["familierelasjon.bor_sammen"] = createSvar(
                if (java.lang.Boolean.TRUE == barn.borSammenMed.verdi) "familie.barn.true.barn.borsammen.true" else "familie.barn.true.barn.borsammen.false",
                SvarType.LOCALE_TEKST
            )
        }
        return Sporsmal(
            tittel = if (erSystemRegistrert) "familie.barn.true.barn.sporsmal" else "familierelasjon.faktum.lagttil",
            erUtfylt = true,
            felt = listOf(
                Felt(
                    type = Type.SYSTEMDATA_MAP,
                    labelSvarMap = labelSvarMap
                )
            )
        )
    }

    private fun deltBostedSporsmal(barn: JsonAnsvar): Sporsmal {
        val harUtfyltDeltBostedSporsmal = barn.harDeltBosted != null && barn.harDeltBosted.verdi != null
        val harSvartJaDeltBosted = harUtfyltDeltBostedSporsmal && java.lang.Boolean.TRUE == barn.harDeltBosted.verdi
        return Sporsmal(
            tittel = "system.familie.barn.true.barn.deltbosted.sporsmal",
            erUtfylt = harUtfyltDeltBostedSporsmal,
            felt = if (harUtfyltDeltBostedSporsmal) StegUtils.booleanVerdiFelt(
                harSvartJaDeltBosted,
                "system.familie.barn.true.barn.deltbosted.true",
                "system.familie.barn.true.barn.deltbosted.false"
            ) else null
        )
    }

    private fun barneBidragSporsmal(forsorgerplikt: JsonForsorgerplikt?): Sporsmal {
        val erUtfylt = forsorgerplikt?.barnebidrag != null && forsorgerplikt.barnebidrag.verdi != null
        return Sporsmal(
            tittel = "familie.barn.true.barnebidrag.sporsmal",
            erUtfylt = erUtfylt,
            felt = if (erUtfylt) listOf(
                Felt(
                    type = Type.CHECKBOX,
                    svar = createSvar(verdiToTekstKey(forsorgerplikt!!.barnebidrag.verdi), SvarType.LOCALE_TEKST)
                )
            ) else null
        )
    }

    private fun verdiToTekstKey(verdi: Verdi): String {
        val key: String = when (verdi) {
            Verdi.MOTTAR -> "familie.barn.true.barnebidrag.mottar"
            Verdi.BETALER -> "familie.barn.true.barnebidrag.betaler"
            Verdi.BEGGE -> "familie.barn.true.barnebidrag.begge"
            Verdi.INGEN -> "familie.barn.true.barnebidrag.ingen"
            else -> "familie.barn.true.barnebidrag.ingen"
        }
        return key
    }
}
